src/compressionworker.ts

```diff
@@ -0,0 +1,105 @@
+import compress from "./compress";
+import { config } from "./config";
+
+export function createCompressionWorker(
+  envelope: IEnvelope,
+  onMessage?: (e: MessageEvent) => void,
+  onError?: (e: ErrorEvent) => void
+): Worker {
+  let workerUrl = createWorkerUrl(envelope);
+  let worker = new Worker(workerUrl);
+  worker.onmessage = onMessage || null;
+  worker.onerror = onError || null;
+  return worker;
+}
+
+function workerContext() {
+  let workerGlobalScope = self as any;
+  let compress = workerGlobalScope.compress;
+  let config = workerGlobalScope.config;
+  let envelope: IEnvelope = workerGlobalScope.envelope;
+  let nextPayloadEvents: IEvent[] = [];
+  let nextPayloadBytes = 0;
+  let sequence = 0;
+
+  // Edge case: Flag to skip uploading payloads consisting of a single XhrError instrumentation event
+  // This helps us avoid the infinite loop in the case when all requests fail (e.g. dropped internet connection)
+  // Infinite loop comes from sending instrumentation about failing to deliver previous delivery failure instrumentation.
+  let nextPayloadIsSingleXhrErrorEvent: boolean =  false;
+
+  onmessage = (evt: MessageEvent) => {
+    let message = evt.data;
+    switch (message.type) {
+      case WorkerMessageType.AddEvent:
+        let addEventMsg = message as IAddEventMessage;
+        addEvent(addEventMsg.event, addEventMsg.time);
+        break;
+      case WorkerMessageType.ForceUpload:
+        let forceUploadMsg = message as ITimestampedWorkerMessage;
+        uploadNextPayload(forceUploadMsg.time);
+        break;
+      default:
+        break;
+    }
+  };
+
+  function addEvent(event: IEvent, time: number): void {
+    let eventStr = JSON.stringify(event);
+
+    // If appending new event to next payload would exceed batch limit, then flush next payload first
+    if (nextPayloadBytes > 0 && nextPayloadBytes + eventStr.length > config.batchLimit) {
+      uploadNextPayload(time);
+    }
```

Comment:
Will we ever stop uploading? It's not clear from this logic it will do so. Better would be to add a state to indicate "stop uploading" and set that once and use here to bail out quickly.

Reply:
Why would we not stop uploading? We are uploading at most 1 batch here. I think i should rename these functions to 'sendToMain' or something like that to avoid confusion, since these calls aren't doing the actual uploading.

---
In reply to: [137238671](https://github.com/Microsoft/clarity-js/pull/28#discussion_r137238671) [](ancestors = 137238671)

Reply:
Renamed these functions and updated comments

---
In reply to: [137354749](https://github.com/Microsoft/clarity-js/pull/28#discussion_r137354749) [](ancestors = 137354749,137238671)